### ISSUE TITLE
_get_column() change proposal

### ISSUE THREAD

**Issue by WoutGeeurickx** [twopirllc/pandas-ta#137]
Hi Kevin,

I noticed some behavior which seemed confusing, at least to me. While using df.ta.hlc3(append=True), I noticed that the column that was created was exactly the 'close' column (mind that I have set df.ta.adjusted = close). Stepping through the code uncovered that this is because I didn't give explicit 'high', 'low' and 'close' columns, and if those are left to None, _get_column()  will return 'adjusted' ('close' in my case) unless that's not set, in which case it will return default which is passed as an argument.

line 388-389 in core.py

```python
    elif series is None or default is None:
        return df[self.adjusted] if self.adjusted is not None else df[default]
```

What I found strange about this implementation is that even if a default is given ('high', 'low', 'close' respectively for hlc3), the function will still return the 'adjusted' column if it is set explicitely. To me that doesn't really make sense with the 'default' name of the argument.

I propose to change the _get_column() function along these lines:

```python
    def _get_column(self, series, default):

        """Attempts to get the correct series or 'column' and return it."""
        df = self._df
        if df is None: return

        # def _get_case(column: str):
        #     cases = [column.lower(), column.upper(), column.title()]
        #     return [c for i, c in enumerate(cases) if column == cases[i]].pop()
        # default = _get_case(default)

        # Explicitly passing a pd.Series to override default.
        if isinstance(series, pd.Series):
            return series
        **# Apply default if no series, or adjusted if no default is passed.
        elif series is None:
            if default is not None:
                series = default
            else:
                series = self.adjusted

        # Try to interpret passed string as a column.
        if isinstance(series, str):**
            # Return the df column since it's in there.
            if series in df.columns:
                return df[series]
            else:
                # Attempt to match the 'series' because it was likely misspelled.
                matches = df.columns.str.match(series, case=False)
                match = [i for i, x in enumerate(matches) if x]
                # If found, awesome.  Return it or return the 'series'.
                cols = ', '.join(list(df.columns))
                NOT_FOUND = f"[X] Ooops!!!: It's {series not in df.columns}, the series '{series}' was not found in {cols}"
                return df.iloc[:,match[0]] if len(match) else print(NOT_FOUND)
```

This way the passed 'default' column gets priority over the 'adjusted' column. Moving the 'if isinstance(series,str)' to a separate 'if' statement also allows for slightly misspelled default column names to work as well.

Since this is a core function I wanted to first open a discussion before sending a PR as it may potentially break things that I'm not aware of, or there may be a particular reason why this has been implemented as it is right now. If that's the case, I'm happy to hear the reasoning as it may help me pay attention to usage of the _get_column() function in the future.

EDIT: cleaned up code markup

Best regards,
Wout

**Comment by twopirllc** [comment]
Hello @WoutGeeurickx,

Thanks for using **Pandas TA**!

Thanks for bringing this into a discussion before making a PR request on ```_get_column(**kwargs)```. The _core_, and namely the ```strategy``` method has gotten quite the work out this year from bugs to enhancements. So it's fitting that ```_get_column(**kwargs)``` gets some enhancement love as well.

From a brief test of the code you provided above would require each indicator to be changed like so:
```python
    # From
    def hlc3(self, offset=None, **kwargs):
        high  = self._get_column(kwargs.pop("high", "high"))
        low   = self._get_column(kwargs.pop("low", "low"))
        close = self._get_column(kwargs.pop("close", "close"))
        ...

    # To (Correct?)
    def hlc3(self, offset=None, **kwargs):
        high  = self._get_column(kwargs.pop("high", "high"), "high")
        low   = self._get_column(kwargs.pop("low", "low"), "low")
        close = self._get_column(kwargs.pop("close", "close"), "close")
        ...
```
But that's an inconsequential change that search and replace can easily handle.

While that may solve the problem, my question is: Can it be done better? Is using ```_get_column(**kwargs)``` the proper solution? Or should the method definition be more explicit and ```_get_column(**kwargs)``` gets a makeover as well.
```python
    def hlc3(self, high="high", low="low", close="close", offset=None, **kwargs):
        high  = self._get_column(high, "high")
        low   = self._get_column(low, "low")
        close = self._get_column(close, "close")
        ...
```

In short, I would like to make ```_get_column(**kwargs)``` more flexible **while** simplifying the core more. Since you have some familiarity of ```_get_column(**kwargs)```, perhaps you may have some other ideas, solutions and improvements? Currently I have additional enhancements pending that I want to get onto the library so my bandwidth for this topic is rather limited unfortunately. So I would appreciate fresh eyes on this topic and welcome to contributions.

Thanks,
Kevin

**Comment by WoutGeeurickx** [comment]
Hi Kevin,

I'm personally more a fan of the more explicit method definition as it helps to understand at a quick glance what the relevant columns are. It seems that I'm locally on an older version, because my hlc3 method still looks like this:

    @finalize
    def hlc3(self, high=None, low=None, close=None, offset=None, **kwargs):
        high = self._get_column(high, 'high')
        low = self._get_column(low, 'low')
        close = self._get_column(close, 'close')

        ...

Strangely I also see now that the _get_column() method no longer has the 'default' argument on the latest Github commit, while I thought I made sure to look at the latest version before posting this proposal. I'm running 0.1.66b locally, so back then the _get_column() method still had the 'default' argument at least. I couldn't tell when this has changed.

Indeed, changing

    high  = self._get_column(kwargs.pop("high", "high"))

to 

    high  = self._get_column(kwargs.pop("high", "high"), "high") 

will give the same functionality if the 'default' argument is added to _get_column() method again.

Probably there's a reason why the 'default' argument got removed over time. I'm also not arguing to get the 'default' argument back if there's no need for it. Maybe even just adding a warning to highlight that the column falls back to 'adjusted' in case series == None could already be an improvement that helps to prevent unexpected behavior.

I currently see three viable options:

1. A warning is generated when the 'series' argument in _get_columns() is None, to let the user know that it will fall back to 'adjusted'. Since there will be no indication of which call to _get_columns() generated this warning, a user will still have to step through the code to rectify this fallback if (s)he desires to do so. But at least they know that it's happening and may prevent unexpected behavior. If the _get_columns() method is called often with series left (intentionally) None, this may clutter the output though.

2. All indicator methods are made explicit again, **but each column argument needs to be defined** (contrary to leaving at as None by default). This will make users think about how the methods are used, but will also require a lot of manual input where this is not necessary with current implementation.

3. The default argument in the _get_column() method is restored and the implementation is changed as described in my original post compared to how it's done in 0.1.66b. This prevents the manual input from Option 2., while still giving (what I believe) more predictable behavior when the columns are not explicitly passed but a default has been set correctly. This can also be complemented with Option 1. in the sense that a warning can be generated if no 'default' argument is passed and the 'adjusted' column is returned as fallback.

I believe from my argumentation it's already clear which option I favor, I'm hoping others can jump in to give some feedback on which option they like most and if I'm maybe missing options or benefits/issues with one of these options.

Best regards,
Wout

**Comment by twopirllc** [comment]
Hello @WoutGeeurickx 

> I'm personally more a fan of the more explicit method definition as it helps to understand at a quick glance what the relevant columns are.

Same here. That is true for when using help on a specific indicator, like ```help(ta.macd)``` et al.

> It seems that I'm locally on an older version, because ...

Yes. Please upgrade to the latest version so that we are on the same page.

> Probably there's a reason why the 'default' argument got removed over time.

Yes. It was done while updating the ```ta.strategy``` method so @DrPaprikaa and I could get _multiprocessing_ working for all the _Strategy_ cases as well as reduce the complexity of defining a _Strategy_. We reduced complexity by adding additional kwargs: ```params``` and ```col_names``` for use in a _Strategy_. See the last example _Additional Strategy Options_ in the [Pandas TA Strategy Examples Notebook](https://github.com/twopirllc/pandas-ta/blob/master/examples/PandasTA_Strategy_Examples.ipynb)

Thanks for the providing some options for consideration. Perhaps they may change when you upgrade and do more tests. Hopefully @DrPaprikaa and others may join in with their ideas and suggestions before moving forward with potentially mass changes.

Thank you,
KJ

**Comment by DrPaprikaa** [comment]
Hi,
If I'm not mistaken this bug is not happening in the current _master_ version. 

`high = self._get_column(kwargs.pop("high", "high"))`

here, `pop` searches for the value associated with the key `high` in kwargs (chained indicator), defaulting with the value `high` (normal indicators). So `_get_column()` is never passed `None`. If `None` is passed for whatever reason, `_get_column()` simply returns `adjusted`, or `None` if not present via the `else None` that I added a few versions back in line 384 : 
```

elif series is None:
    return df[self.adjusted] if self.adjusted is not None else None
```
So I think that nothing needs to be changed, correct me if I'm wrong :)

DrPaprikaa

**Comment by WoutGeeurickx** [comment]
Hi @DrPaprikaa ,

You are absolutely correct! I'm less experienced with the usage of kwargs, so I wasn't aware that the second parameter in kwargs.pop() is the fallback in case the key is not found. This works exactly as what I tried to propose, and in a clean way (although I like the explicit method definition 'of old').
I'm sorry for raising this issue since it seems that everything would just have been solved if I had updated to the latest version. I still find it strange because I made sure to look up that piece of the source code on what I believed what was the latest version.

@twopirllc , I believe this issue can be closed, unless of course you want to keep this as a discussion thread for explicit method definition?

**Comment by twopirllc** [comment]
Hey guys,

Thanks for chipping in @DrPaprikaa.

> I'm sorry for raising this issue since it seems that everything would just have been solved if I had updated to the latest version. I still find it strange because I made sure to look up that piece of the source code on what I believed what was the latest version.

No worries.

> @twopirllc , I believe this issue can be closed, unless of course you want to keep this as a discussion thread for explicit method definition?

No need to keep it open if the issue is resolved. It will be under the closed issues if it needs linking for future discussions.

Thanks for having this discussion before making a Pull Request.

Best Regards,
KJ
